### PR TITLE
Memory management for binary separation

### PR DESCRIPTION
**Pull request by kishore-sn**
- Backport MM changes from nuttx
- Fix coding rule violations in previous commit

This is a resubmission of https://github.com/Samsung/TizenRT/pull/2746 after fixing all issues.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [3b4fedc9a5765703ccee90e648470c0bf14cc742] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [3b4fedc9a5765703ccee90e648470c0bf14cc742] - Code Rule Check (C++) OK.

File: os/mm/mm_heap/mm_getheap.c [hunk 1]
```diff
@@ -95,12 +100,11 @@ int mm_get_heapindex(void *mem)
 			/* A valid address from the user heap for this region would have to lie
 			 * between the region's two guard nodes.
 			 */
-			if ((mem > (void *)g_mmheap[heap_idx].mm_heapstart[region]) && (mem < (void *)g_mmheap[heap_idx].mm_heapend[region])) {
+			if (mem < (void *)(USR_HEAP[heap_idx].mm_heapstart + USR_HEAP[heap_idx].mm_heapsize) && mem >= (void *)USR_HEAP[heap_idx].mm_heapstart) {
```
Comment (jeongchanKim) [inline os/mm/mm_heap/mm_getheap.c]: This is chaged some days ago.
(https://github.com/Samsung/TizenRT/commit/447944f1ed595f7c5119757d28562f8ea9ce15e2#diff-0b5f7261f9d67c06ac44fb1c7ab14d5e)
Do you have any reason for reverting this?

File: os/mm/mm_heap/mm_getheap.c [hunk 2]
```diff
@@ -26,8 +26,17 @@
 #include <tinyara/sched.h>
 #endif
 
-#if !defined(CONFIG_BUILD_PROTECTED) || !defined(__KERNEL__)
+#if defined(CONFIG_BUILD_PROTECTED) && defined(__KERNEL__)
```
[the remark this answers is not in the document]
Comment (kishore-sn) [reply os/mm/mm_heap/mm_getheap.c]: No. USR_HEAP is defined in umm_heap.h, and this file cannot be included in mm_getheap.c.
Hence we are redeclaring it here.

File: os/mm/mm_heap/mm_getheap.c (hunk 1 above)
Comment (kishore-sn) [reply os/mm/mm_heap/mm_getheap.c]: Here, we are just changing g_mmheap to USR_HEAP.
The changes that you mentioned are in master branch and yet to be integrated to protection branch.

**Comment by seinfra** [comment]
Target : [ceada7485308670524567af1a129932e2be73213] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ceada7485308670524567af1a129932e2be73213] - Code Rule Check OK.

**Pull request merged**